Initial creation of enqueue objects

steveburkett opened the pull request:
Create the rake task that reads rex releases for book versions to be indexed.   For each book in the rex release, create a dynamo entry for it and enqueue a SQS job.   

The DynamoDB entry records the book version metadata to be indexed.  

The SQS job is the job to be executed for indexing by a worker.  

https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/197

steveburkett committed:
Initial creation of enqueue objects

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,36 @@
+# Indexes a book version thru the SQS queuing system
+#
+# @params indexing_version e.g., Search::BookVersions::I1::IndexingStrategy::VERSION
+class EnqueueIndexJobs
```
My reading of this class was that it contained all the work for the rake script, so the rake script could just call off to this PORO.  But looking at how it is initialized with a single indexing strategy I'm getting the feeling that may not be the case?

If this is intended to be just for one indexing strategy, that means that when running for multiple indexing strategies we'll be repulling and iterating through all the releases multiple times.  I'm not sure if that is the best.  

Above is discussed how the validation approach to checking if a BookIndexing already exists might not be ideal.  Also with our desire to delete no-longer-requested book/indexing strategy combos, an all-at-once approach like the following might be better (the below is very rough):

```ruby
class EnqueueIndexJobs
  def do_it
    rex_releases = OpenStax::RexReleases.new
    book_ids = rex_releases.map{ |release| release.books }.flatten

    book_indexings = BookIndexing.all # guessing that this is how you get all entries from Dynamo

    book_ids.each do |book_id|
      active_indexing_strategies.each do |indexing_strategy|
        # to avoid iterating through the entire book_indexings array over and over, probably good
        # to make a hash of book IDs to hashes of indexing strategies to BookIndexings

        book_indexing = that_crazy_hash[book_id].try(:[], indexing_strategy)

        if book_indexing.nil?
          # hey, we haven't done this one yet! 
          job = IndexingJob.new(book_version_id: ..., indexing_strategy: ...)
          @todo_jobs_queue.write(job) 
          BookIndexing.create!(...)
        else
          # we have done this one already, somehow note that it is still in demand
          # ones we don't mark don't have a book ID / indexing strategy combination
          # indicated in the released content, and should be deleted (below)
        end
    end

    # find all of our local book indexings that we didn't mark in demand and delete 
    # all of their indexes

  end
end
```

jpslav commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,57 @@
+
+  def book_indexing_version_is_unique
+    found_books = BookIndexing.where(book_version_id: book_version_id,
+                                     indexing_version: indexing_version)
```
I'm pretty sure this is not an efficient way to operate.  DynamoDB is a simple key value store and can't do efficient queries like a RDBMS.  Looking at the original Dynamoid blog post, a `where` query "does not efficiently perform a conditional join, but instead finds all the user's addresses and naively filters them in Ruby".  So combined with the loop below in `EnqueueIndexJobs` this means we are querying all of the book indexing records over and over.  I suspect we'll want to find a different approach.

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,36 @@
+    book_indexing.destroy if book_indexing
+
+    if ex.is_a?(Dynamoid::Errors::DocumentNotValid)
+      Rails.logger.info "Book version #{book_id} already being indexed by IndexingJob"
```
In fact it may have been indexed weeks ago, so we don't want to log this (would be a lot of duplicated log messages for what is an expected situation)

jpslav reviewed (changes requested):
I did not review the specs but I do like that there are a lot of them!

jpslav committed:
making enqueue code more inline

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,70 @@
+      end
+    end
+
+    check_for_books_to_delete
```
I understand the desire to put code into private methods, but here it doesn't help me read this code and understand what's going on.  "process_book" is kind of vague.  Also, `process_book` sets state that is used in `check_for_books_to_delete` so for that I'd rather just have the code more inline here.  I made a PR into this PR to show you what I mean - it ends up as 24 lines of code but keeps logic together and is more readable to me (note that that PR is pseudocode-y so you may not just be able to merge it).

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,70 @@
+
+    check_for_books_to_delete
+
+    AutoScaling.set_desired_capacity(group_name: asg_name, desired_capacity: @todo_jobs_queue.count)
```
In that PR I changed this to just an ASG instance (more instance methods, less class methods)

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,70 @@
+  def find_book_in_indexing(book_id, indexing_version)
+    book_indexings.detect do |book_indexing|
+      book_indexing.book_version_id == book_id &&
+         book_indexing.indexing_version == indexing_version
```
This could be pretty inefficient -- iterating through 1500 entries for each of 1500 combinations of inputs; O(N^2).  In my PR I cached a hash of the book indexings (O(N)) to make lookup O(1).

steveburkett replied:
this iterates thru the book indexings just once and then memoizes the results.

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,70 @@
+    BookIndexing.create(book_version_id: book_id, indexing_version: indexing_version, state: state)
+
+    Rails.logger.info "Open-Search: Book version #{book_id} #{indexing_version} loaded into todo_job_queue"
+  end
```
in PR switched these to `enqueue_create_index_job` and `enqueue_delete_index_job`.  Even if don't use these words, prefer "enqueing jobs" rather than "enqueueing books" b/c we aren't really queuing books.

jpslav replied:
this old review comment appears to be disconnected from its original location?

jpslav commented on config/initializers/indexing_strategies.rb:
```diff
@@ -0,0 +1,7 @@
+ACTIVE_INDEXING_STRATEGIES = [
+  Search::BookVersions::I1::IndexingStrategy
```
could put the strings here instead of classes, so don't have to say `.new.version` above which may not be same for all strategies

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,70 @@
+    @todo_jobs_queue = TodoJobsQueue.new
+  end
+
+  def process_rex_releases
```
changed name to `execute` in the PR.  B/c does an outside caller really care that this starts by pulling rex releases?  I think this Ruby PORO is still just a script but now one that has attributes/state.  So I like just saying `do_it` or `call` or similar.

steveburkett replied:
this is just my opinion...so would love to discuss more...but i've disliked using "meaningless" method names like execute or do_it or call (in services).  it adds nothing to the code documentation and understanding, when what is it "calling" or "doing' is adding to easier code understanding.  method names are the source for documenting what the code or method is doing, hopefully a single responsibility.

jpslav replied:
this method is what could be the inlined body of the rake script, so to me the name of this method is really "enqueue_index_jobs".  of course that is also the class name here, which is why a "do_it" method makes sense to me (because `EnqueueIndexJobs.new.enqueue_index_jobs` is repetitive).  Said another way, `EnqueueIndexJobs` is a glorified function (a function that is an object so that it can have state), and functions are `call`ed.  Does `process_rex_releases` help here with the understanding of what's going on?  If a new user came to this code, would it be obvious that `process_rex_releases` is what should be called to run this "enqueue index jobs" function?  To me, it isn't really obvious other than it being the only public method.

steveburkett replied:
thanks @jpslav.  We argued over this for hours at my previous job.... i'm cool with this and agree...that the idea of the service object is that of a glorified function object, to invoke call or execute on. 

i do hope tho that we dont pack our (production) rake scripts with functional code.  same reasoning as packing a controller with functional code (harder to test, less SRP, etc).

jpslav commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,59 @@
+class BookIndexing
+  include Dynamoid::Document
+
+  VALID_STATES = [
```
These are all the states, so why prefix with "VALID_".  Is there an "invalid" state?  Maybe just call "STATES"

jpslav commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+  end
+
+  def self.valid_book_indexings
```
This use of "valid" doesn't match "VALID_STATES" above.  Also, I don't think a "deleted" state is invalid per se.

steveburkett replied:
@jpslav do you like the states prefixed like this? 
```
  STATES = [
    STATE_PENDING = "pending",
    STATE_DELETE_PENDING = "delete pending",
    STATE_STARTED = "started",
    STATE_INDEXED = "indexed",
    STATE_DELETED = "deleted"
  ]
```

jpslav commented on app/models/todo_jobs_queue.rb:
```diff
@@ -0,0 +1,14 @@
+
+  def read
+    message = raw_queue.receive_messages()
+    message.size == 0 ? nil : IndexingJob.new(message)
```
You can comment this out until working on worker code if you want

steveburkett commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,70 @@
+    end
+
+    books_to_be_deleted.each do ||
+      job = DeletingJob.new(book_version_id: book_id, indexing_version: indexing_version)
```
yeah...agreed.  i thought the naming of this value object could be better too....

steveburkett committed:
Merge pull request #23 from openstax/jp/schedule_indexing_jobs-1

making enqueue code more inline

steveburkett committed:
Cleanup for pull request review and get working again

steveburkett commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,72 @@
+  STATES = [
+    STATE_PENDING = "pending",
+    STATE_DELETE_PENDING = "delete pending",
+    STATE_STARTED = "started",
```
@jpslav i just guessed at this...i think it's what we'd set the dynamo object to when this job is started.  does that sound right?   (and then indexed when finished).

steveburkett commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,72 @@
+class BookIndexing
```
if we rename this, my preference would be something like BookIndexingJob or the like.  because it mostly accurately represents a job (in the domain world)...it contains state and a few other things (like timestamps).  

i do agree we should strive toward the best naming we can muster here...the word indexing is kind overloaded.

steveburkett replied:
@jpslav and I chatted more on this.   it was a miscommunication on my part that this class should be a state of indexing, and not a "activity log".  so...will rename to JP's recommendation.

steveburkett commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,87 @@
+
+  def call
+    released_book_ids.each do |book_id|
+      ACTIVE_INDEXING_STRATEGIES.each do |indexing_version|
```
i used indexing_version because that what it was.   a strategy is the indexing object/class.  agree we should be consistent.  are u ok with calling this ACTIVE_INDEXING_VERSIONS?

steveburkett committed:
Attempts to make specs work on travis

steveburkett committed:
Simply DynamoidReset

steveburkett committed:
Use Aws SDK to create indexing table for travis

steveburkett commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,72 @@
+  attr_reader :in_demand
+
+  def self.create_new_indexing(book_version_id:, indexing_version:)
+    Rails.logger.info "Creating book version #{@book_version_id} #{@indexing_version} in dynamodb"
```
i can use less logging.  I'm used to a shop where we logged everything (because it was hard to figure out what was wrong in the system).

steveburkett commented on app/models/book_indexing.rb:
```diff
@@ -0,0 +1,72 @@
+    self.state = STATE_DELETE_PENDING
+    self.enqueued_time = DateTime.now
+    save!
+  end
```
absolutely!  good catch.

jpslav committed:
fool the AWS checksum check in to thinking it passes during VCR playback; fix dynamo table name mismatch

steveburkett commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,87 @@
+                          indexing_version: book_indexing.indexing_version)
+    @todo_jobs_queue.write(job)
+
+    book_indexing.queue_to_delete
```
mark_queued_for_deletion is a more accurate name

steveburkett commented on spec/support/sqs.rb:
```diff
@@ -0,0 +1,17 @@
+module Sqs
+  def self.create_test_queue
+    random_queue_name = "test-#{(0...8).map { (65 + rand(26)).chr }.join}"
```
using a simple constant test queue name...dont see the need for a random name....

steveburkett committed:
Fix comments made in code review #2

steveburkett commented on spec/models/book_indexing_spec.rb:
```diff
@@ -1,9 +1,13 @@
 require 'rails_helper'
 require 'vcr_helper'
 
+amazon_api_header_matcher = lambda do |request_1, request_2|
```
great hackery!

jpslav committed:
fix specs by aligning secrets.yml.example to cassettes

jpslav commented on app/models/book_index_state.rb:
```diff
@@ -0,0 +1,74 @@
+
+  def mark_queued_for_deletion
+    self.state = STATE_DELETE_PENDING
+    self.enqueued_time = DateTime.now
```
What time zone does this get set in?  Brief looking around in Dynamoid suggests server's local time. Should we standardize to UTC?

steveburkett replied:
for sure we should use UTC (will test this now to confirm what dynamoid stores)

steveburkett replied:
yeah, dynamoid saves as UTC time

jpslav commented on app/models/book_index_state.rb:
```diff
@@ -0,0 +1,74 @@
+
+  def self.live_book_indexings
+    all.reject{ |doc| doc.deleting? }
+  end
```
Change name away from `book_indexings` to match name change of class.  Also here `live` could work

jpslav commented on app/services/enqueue_index_jobs.rb:
```diff
@@ -0,0 +1,94 @@
+
+  def book_indexings
+    @book_indexings ||= BookIndexState.live_book_indexings
+  end
```
Change name of this method/variable and other uses in this class away from "BookIndexing" to match name change to BookIndexState, e.g. `def index_states`

steveburkett commented on app/models/book_index_state.rb:
```diff
@@ -0,0 +1,74 @@
+  field :started_time,  :datetime, store_as_string: true
+  field :finished_time, :datetime, store_as_string: true
+  field :updated_at,    :datetime, store_as_string: true
+  field :created_at,    :datetime, store_as_string: true
```
@jpslav i put these here so we could get store_as_string = true instead of the unix time, which is easier to grok

steveburkett commented on app/models/book_index_state.rb:
```diff
@@ -0,0 +1,74 @@
+  field :state
+  field :enqueued_time, :datetime, store_as_string: true
+  field :started_time,  :datetime, store_as_string: true
+  field :finished_time, :datetime, store_as_string: true
```
in #198, we'll move to a status_log hash, replacing the timestamps

steveburkett committed:
Fix comments in code review #3

Main points:
(1) not using book_indexing moniker (using states, or live)
(2) we'll move away from using timestamps in columns to represent
    manager work.  will use a status log hash instead (later)
